File: MML-Card-BE/Models/Card/Useable/Ability.cs
using MML_Card_BE.Models.Enum;
using MongoDB.Bson.Serialization.Attributes;
using MongoDB.Bson;
using MML_Card_BE.Models.Card.Useable.Interfaces;

namespace MML_Card_BE.Models.Card.Useable
{
    public class Ability : IAbility
    {
        [BsonId]
        [BsonRepresentation(BsonType.ObjectId)]
        public string? _id { get; set; }
        [BsonElement("name")]
        public string Name { get; set; }

        [BsonElement("onUse")]
        [BsonRepresentation(BsonType.String)]
        public Func<string> OnUse { get; set; }
    }
}

File: client/src/models/Creature/Ability.ts
interface IAbility {
  _id: number;
  name: string;
  onUse: Function;
  description: string;
}

class Ability implements IAbility {
  _id: number;
  name: string;
  description: string;
  onUse: Function;

  constructor(_id: number, name: string, description: string, onUse: Function) {
    this._id = _id;
    this.name = name;
    this.description = description;
    this.onUse = onUse;
  }
}

function createGenericAbility(): Ability {
  return new Ability(
    0, // _id
    'Generic Ability Card', // name
    'This is the Ability Description', // description
    () => {
      console.log('This is the onUse function');
    }
  );
}

export default Ability;
